### ISSUE TITLE
Mixed HTTP/WebSocket server?

### ISSUE THREAD

**Issue by nickray** [paritytech/jsonrpsee#821]
Is it correct that currently `http-server` and `ws-server` exclusively serve over HTTP and WebSocket, respectively?

We have seen the desire to have a single socket that can respond to HTTP requests, but also upgrade to WebSocket if requested.

**Comment by niklasad1** [comment]
right, I guess @nickray wants to have a WS/HTTP server on the same socket/server which makes sense.

I reckon that should be possible to something "similar"  https://github.com/paritytech/soketto/blob/master/examples/hyper_server.rs, actix does probably something similar under the hood because it supports both WS/HTTP IIRC.

**Comment by niklasad1** [comment]
As we are already have all code how to do that we hyper so it should just be matter to extract that code into some shared place in such as `core`. Then enable it via a feature flag and/or builder option.

sounds like a good idea too me

**Comment by nickray** [comment]
Yes the interpretation is correct, most clients would like to use websockets to lower the overhead of calling my server by keeping the connection open. However, some use a simpler client that doesn't have the websocket logic, or simply want to do quick curl debugs.
